### PR TITLE
[feat] custom load dependencies

### LINKED ISSUE #1684: (re)trigger the load function
**Is your feature request related to a problem? Please describe.**
The new `invalidate(href)` feature in `$app/navigation` is a great way to trigger the execution of a `load` function that works in most cases. The problem is that this only works if one uses the `fetch` function that is provided as an argument to `load`. If a apiClientWrapper is used this does not work anymore. 

**Describe the solution you'd like**
One solution would be to pass a `invalidate` or `update` function to the `load` so that I can be passed as a prop or via context.
This solution would update a specific `load` function instead of a specific fetch url.

```svelte
<script context="module">
import { api } from '$lib/myapi';
export async function load({page, update}){
    const item = await api.getItem(page.params.id);
    return { 
        props: { item, update }
    }
}
</script>

<script>
import Item from '$lib/Item.svelte';
export let item;
export let update;
</script>

<Item {item} />
<button on:click={update}>Reload</button>
```

**Describe alternatives you've considered**
A alternative would be to write a second function in the (non module) script tags that replicates the load function to update data.

**How important is this feature to you?**
Somewhat important. It would make the code cleaner and more reusable but there are easy workarounds.

### PR DESCRIPTION
**Pull request by icalvin102**
Closes #1684 

### Problem
The `invalidate` function only re-runs `load` if fetch is used. This is not possible in cases where an `APIClient` does not allow for customization of the fetching process or when using non REST APIs like GraphQL.

### Solution
Add a `dependencies` array to the `LoadOutput` (suggested by @Rich-Harris in https://github.com/sveltejs/kit/pull/1709).
The dependencies will be registered the in the same way `fetch` does it. This allows not only to `invalidate` real resources but entire pages or groups of pages.

**Example Usage**

```svelte
<script context="module">
import { api } from '$lib/myapi';
export async function load({params}){
    const item = await api.getItem(params.id);
    return { 
        props: { item },
        dependencies: [`/item-${params.id}`, 'custom:item-page']
    }
}
</script>

<script>
import Item from '$lib/Item.svelte';
import { invalidate } from '$app/navigation';
export let item;
</script>

<Item {item} />
<button on:click={()=> invalidate('custom:item-page') }>Reload Page</button>
<button on:click={()=> invalidate(`/item-${item.id}`) }>Reload Item</button>
```

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [X] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [X] This message body should clearly illustrate what problems it solves.
- [X] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [X] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [X] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpx changeset` and following the prompts. All changesets should be `patch` until SvelteKit 1.0

### CODE REVIEW

**Comment by dreitzner** [comment]
@icalvin102 just a quick question.... 
Is this problem not solved by [page endpoints](https://kit.svelte.dev/docs/routing#endpoints-page-endpoints)?
To the best of my understanding this should generate a load function on build.
By moving the `$lib/myapi` integration to the endpoint there should be no trouble on calling the load again ;)

**Comment by icalvin102** [comment]
@dreitzner page endpoints will only work if your API is written in sveltekit AFAIK.
Page endpoints wont work in SPA mode and/or when using third party APIs that rely on browser state like cookies that are not accessible from the sveltekit server.

File: documentation/docs/04-loading.md [hunk 1]
```diff
@@ -134,3 +134,11 @@ If the `load` function returns a `props` object, the props will be passed to the
...
 
 The combined `stuff` is available to components using the [page store](/docs/modules#$app-stores) as `$page.stuff`, providing a mechanism for pages to pass data 'upward' to layouts.
+
+#### dependencies
+
+If the `load` function depends on custom resources a `dependencies` array can returned. A registered dependency can later be used as argument of `invalidate` to rerun the `load` function.
+
+> Note that the `fetch` function alreay registers the requested resources as dependencies. Manual registration is only necessary if it's not possible to use SvelteKits `fetch` function. 
+
+The items should be [properly encoded](https://developer.mozilla.org/en-US/docs/Glossary/percent-encoding) URIs. Both absolute and relative URIs are acceptable.
```
Comment (Rich-Harris) [inline documentation/docs/04-loading.md:144]:
```suggestion
An array of strings representing URLs the page depends on, which can subsequently be used with [`invalidate`](/docs/modules#$app-navigation-invalidate) to cause `load` to rerun. You only need to add them to `dependencies` if you're using a custom API client; URLs loaded with the provided `fetch` function are added automatically.

URLs can be absolute or relative to the page being loaded, and must be [encoded](https://developer.mozilla.org/en-US/docs/Glossary/percent-encoding).
```

**Review by Rich-Harris — APPROVED** [review]

**Comment by Rich-Harris** [comment]
thank you!

**Pull request merged**